backends/cuda/ceed-cuda-operator.c

```diff
@@ -729,51 +729,53 @@ __device__ void diagonalCore(const CeedInt nelem,
-        // Each component
-        for (CeedInt compOut = 0; compOut < NCOMP; compOut++) {
-          // Each qpoint/node pair
-          if (pointBlock) {
-            // Point Block Diagonal
-            for (CeedInt compIn = 0; compIn < NCOMP; compIn++) {
-              CeedScalar evalue = 0.;
-              for (CeedInt q = 0; q < NQPTS; q++) {
+  CeedInt dout = -1;
+  // Each basis eval mode pair
+  for (CeedInt eout = 0; eout < NUMEMODEOUT; eout++) {
+    const CeedScalar *bt = NULL;
+    if (emodeout[eout] == CEED_EVAL_GRAD)
+      dout += 1;
+    CeedOperatorGetBasisPointer_Cuda(&bt, emodeout[eout], identity, interpout,
+                                     &gradout[dout*NQPTS*NNODES]);
+    CeedInt din = -1;
+    for (CeedInt ein = 0; ein < NUMEMODEIN; ein++) {
+      const CeedScalar *b = NULL;
+      if (emodein[ein] == CEED_EVAL_GRAD)
+        din += 1;
+      CeedOperatorGetBasisPointer_Cuda(&b, emodein[ein], identity, interpin,
+                                       &gradin[din*NQPTS*NNODES]);
+      // Each component
+      for (CeedInt compOut = 0; compOut < NCOMP; compOut++) {
+        // Each qpoint/node pair
+        if (pointBlock) {
+          // Point Block Diagonal
+          for (CeedInt compIn = 0; compIn < NCOMP; compIn++) {
+            for (CeedInt q = 0; q < NQPTS; q++) {
+              const CeedScalar bt_q = bt[q*NNODES+tid], b_q = b[q*NNODES+tid];
+              // Each element
+              for (CeedInt e = blockIdx.x*blockDim.z + threadIdx.z; e < nelem;
+                  e += gridDim.x*blockDim.z) {
                 const CeedScalar qfvalue =
                   assembledqfarray[((((ein*NCOMP+compIn)*NUMEMODEOUT+eout)*
                                      NCOMP+compOut)*nelem+e)*NQPTS+q];
                 if (abs(qfvalue) > qfvaluebound)
-                  evalue += bt[q*NNODES+tid] * qfvalue * b[q*NNODES+tid];
+                  elemdiagarray[((compOut*NCOMP+compIn)*nelem+e)*NNODES+tid] += bt_q * qfvalue * b_q;
               }
-              elemdiagarray[((compOut*NCOMP+compIn)*nelem+e)*NNODES+tid] += evalue;
             }
-          } else {
-            // Diagonal Only
-            CeedScalar evalue = 0.;
-            for (CeedInt q = 0; q < NQPTS; q++) {
+          }
+        } else {
+          // Diagonal Only
+          for (CeedInt q = 0; q < NQPTS; q++) {
+            const CeedScalar bt_q = bt[q*NNODES+tid], b_q = b[q*NNODES+tid];
+            // Each element
+            for (CeedInt e = blockIdx.x*blockDim.z + threadIdx.z; e < nelem;
+                e += gridDim.x*blockDim.z) {
               const CeedScalar qfvalue =
                 assembledqfarray[((((ein*NCOMP+compOut)*NUMEMODEOUT+eout)*
                                    NCOMP+compOut)*nelem+e)*NQPTS+q];
               if (abs(qfvalue) > qfvaluebound)
-                evalue += bt[q*NNODES+tid] * qfvalue * b[q*NNODES+tid];
+                elemdiagarray[(compOut*nelem+e)*NNODES+tid] += bt_q * qfvalue * b_q;
```

Comment:
I'm not sure why this inner loop over elements is good memory locality. And do we need the conditional on `qfvaluebound`? I'd expect it actually prevents the read/store from being executed on all lanes, which would require higher occupancy to hide. But each thread processes all the quadrature points so higher occupancy probably isn't possible. (Do you have profiler/occupancy stats?)

I'd sort of expect that a good implementation would load `assembledqfarray` into shared memory (single element or small batch per thread block) and have one thread per (component, node). Tuning such a thing might require specializing on sizes, which we'd rather not do here.

Reply:
I made the switch pulling 2 basis entries from memory is more costly than pulling 1 entry from the assembledqfarray.

I could try to drop the conditional. I originally added that check because repeatedly adding nearly zero values added a ton of error.

Ultimately this whole thing needs to be rebuilt in the style of cuda-gen to actually work well, where we build assembledqfarray as we need it, but that's a bigger PR and I only want to use this approach in the cuda-gen backend.

Reply:
If it's an accuracy issue, I think it would be better to preprocess inputs so values that should be exactly zero actually are.